kernel/src/platform/system_timer.rs

```diff
@@ -0,0 +1,139 @@
+//! Interface system tick timer.
+
+use crate::hil::time::{self, Frequency};
+
+/// Interface for the system tick timer.
+///
+/// A system tick timer provides a countdown timer to enforce process scheduling
+/// quantums.  Implementations should have consistent timing while the CPU is
+/// active, but need not operate during sleep.
+///
+/// On most chips, this will be implemented by the core (e.g. the ARM core systick), but
+/// some chips lack this optional peripheral, in which case it might be
+/// implemented by another timer or alarm controller, or require virtualization
+/// on top of a single hardware timer.
+pub trait SystemTimer {
```

Comment:
```suggestion
pub trait TimesliceAlarm {
```
Maybe? It seems we should at least be consistent with the HIL in regards to alarm vs. timer. This alarm is a bit strange in that it doesn't have a client, so it's largely not useful except for in the core kernel loop that can check for `expired()`, so something specific like "timeslice" or "scheduling" makes sense to me.

Reply:
I agree with Brad here. I would go further and say that we should try to use the word `alarm` throughout the comments, documentation, and filenames as well for this, as in Tock semantics, this is acting as an `alarm` (never mind what ARM happened to call it).

Reply:
From `kernel/src/hil/time.rs`:
```rust
/// The `Alarm` trait models a wrapping counter capable of notifying when the
/// counter reaches a certain value.

//...

/// The `Timer` trait models a timer that can notify when a particular interval
/// has elapsed.
```

Given that the exposed interface is "fire an interrupt x us in the future" I think that it is in fact a timer according to those definitions?

Reply:
Haha well I'm not going to argue that Tock isn't stingy with documentation. But that should read "every time a particular interval has elapsed" (I'm pretty sure). It would be nice to see the updated time HIL to confirm.

Reply:
The trait currently specifies oneshot and repeating timers. 

```rust
/// The `Timer` trait models a timer that can notify when a particular interval
/// has elapsed.
pub trait Timer<'a, W = u32>: Time<W> {
    /// Set the client for interrupt events.
    fn set_client(&'a self, client: &'a dyn TimerClient);

    /// Sets a one-shot timer to fire in `interval` clock-tics.
    ///
    /// Calling this method will override any existing oneshot or repeating timer.
    fn oneshot(&self, interval: W);

    /// Sets repeating timer to fire every `interval` clock-tics.
    ///
    /// Calling this method will override any existing oneshot or repeating timer.
    fn repeat(&self, interval: W);

    /// Returns the interval for a repeating timer.
    ///
    /// Returns `None` if the timer is disabled or in oneshot mode and `Some(interval)` if it is
    /// repeating.
    fn interval(&self) -> Option<W>;
```

cc @phil-levis thoughts on if this should be called a timer or an alarm?

Reply:
Timers allow you to precisely control the interval, but not the "when" (because there may be latency before the interval is put into a register). In that way, this looks more like a timer.  I'd suggest making its method names more closely aligned to timer. E.g., we probably don't want to say "interrupts" in the trait.

Can someone explain the division between HIL traits and platform traits? Should this perhaps be in the time HIL?

Reply:
The watchdog trait is currently in HIL which seems inconsistent with this one being in platform, but I am not sure which should move.

Also, I am not sure what to say other than interrupts..that function call is used only to affect whether an interrupt will fire when the timer hits 0. just having `enable()` and `disable()` would be misleading, because it seems like `disable()` would stop the timer, but all we want it to do in this case is make sure no interrupt fires when the timer expires, without otherwise changing the state of the timer.

I am open to name suggestions though. I could replace `config_interrupts(enabled: bool)` with `enable_interrupts()` and `disable_interrupts()`?

Reply:
I spoke with Phil about a name for `config_interrupts()` today, we settled on `arm()` and `disarm()`.

I am leaning towards `SchedulerTimer` instead of `SystemTimer` but will get final feedback on the call tomorrow before changing all the instances

Reply:
renamed to `SchedulerTimer`, marking this as resolved.